lazy.go

```diff
@@ -69,13 +73,17 @@ func (l *lazyConn) Read(b []byte) (int, error) {
 }
 
 func (l *lazyConn) readHandshake() error {
-	l.rhlock.Lock()
-	defer l.rhlock.Unlock()
-
+	l.rhlock.RLock()
+	rhsync := l.rhsync
+	l.rhlock.RUnlock()
 	// if we've already done this, exit
-	if l.rhsync {
+	if rhsync {
 		return l.rerr
 	}
+
+	l.rhlock.Lock()
+	defer l.rhlock.Unlock()
+
 	l.rhsync = true
```

Comment:
You are right, I think I missed because I restructured this code twice while writing it reverting to simpler solution.

Let's figure out https://github.com/multiformats/go-multistream/pull/17#discussion_r130704764 first.